File: eg/remote/remote.c
/*
     ____ __     ____   ___    ____ __         (((((()
    | |_  \ \  /   ) ) | |  ) | |_  \ \  /  \(@)- /
    |_|__  \_\/  __)_) |_|_/  |_|__  \_\/   /(@)- \
                                               ((())))
 *//**
 *  \file  remote.c
 *  \brief  ev3dev-c example of "brick" simplified library.
 *  \author  Vitaly Kravtsov (in4lio@gmail.com)
 *  \copyright  See the LICENSE file.
 */

#include <stdio.h>
#include "coroutine.h"
#include "brick.h"

#define SPEED_LINEAR    75  /* Motor speed for linear motion, in percents */
#define SPEED_CIRCULAR  50  /* ... for circular motion */

int max_speed;         /* Motor maximal speed (will be detected) */

#define MOTOR_LEFT     OUTC
#define MOTOR_RIGHT    OUTB
#define MOTOR_BOTH     ( MOTOR_LEFT | MOTOR_RIGHT )

#define IR_CHANNEL     0

POOL_T ir;             /* IR sensor port (will be detected) */

enum {
	STOP,
	FORTH,
	BACK,
	LEFT,
	RIGHT,
};
int command = STOP;    /* Command for `drive` coroutine */

int alive;             /* Program is alive */

int init( void )
{
	if ( tacho_is_plugged( MOTOR_BOTH, TACHO_TYPE__NONE_ )) {  /* any type of motor */
		max_speed = tacho_get_max_speed( MOTOR_LEFT, 0 );
		tacho_reset( MOTOR_BOTH );
	} else {
		printf( "Please, plug LEFT motor in C port,\n"
		"RIGHT motor in B port and try again.\n"
		);
		/* Inoperative without motors */
		return ( 0 );
	}

	ir = sensor_search( LEGO_EV3_IR );
	if ( ir ) {
		ir_set_mode_ir_remote( ir );

		printf(	"IR remote control:\n"
		"RED UP   & BLUE UP   - forward\n"
		"RED DOWN & BLUE DOWN - backward\n"
		"RED UP   | BLUE DOWN - left\n"
		"RED DOWN | BLUE UP   - right\n"
		);
	} else {
		printf( "IR sensor is NOT found.\n"
		"Please, use the EV3 brick buttons.\n"
		);
	}
	printf( "Press BACK on the EV3 brick for EXIT...\n" );

	return ( 1 );
}

CORO_CONTEXT( handle_ir_control );
CORO_CONTEXT( handle_brick_control );
CORO_CONTEXT( drive );

/* Coroutine of IR remote control handling */
CORO_DEFINE( handle_ir_control )
{
	CORO_LOCAL int btns, pressed = IR_REMOTE__NONE_;

	CORO_BEGIN();
	if ( ir == SOCKET__NONE_ ) CORO_QUIT();

	for ( ; ; ) {
		/* Waiting any IR RC button is pressed or released */
		CORO_WAIT(( btns = sensor_get_value( IR_CHANNEL, ir, IR_REMOTE__NONE_ )) != pressed );
		pressed = btns;

		switch ( pressed ) {
		/* Forward */
		case RED_UP_BLUE_UP:
			command = FORTH;
			break;
		/* Backward */
		case RED_DOWN_BLUE_DOWN:
			command = BACK;
			break;
		/* Left */
		case RED_UP:
		case RED_UP_BLUE_DOWN:
		case BLUE_DOWN:
			command = LEFT;
			break;
		/* Right */
		case BLUE_UP:
		case RED_DOWN_BLUE_UP:
		case RED_DOWN:
			command = RIGHT;
			break;
		/* Stop */
		case IR_REMOTE__NONE_:
		case RED_UP_RED_DOWN:
		case BLUE_UP_BLUE_DOWN:
		case BEACON_MODE_ON:
			command = STOP;
			break;
		}
		CORO_YIELD();
	}
	CORO_END();
}

/* Coroutine of the EV3 brick keys handling */
CORO_DEFINE( handle_brick_control )
{
	CORO_LOCAL uint8_t keys, pressed = EV3_KEY__NONE_;

	CORO_BEGIN();
	for ( ; ; ) {
		/* Waiting any brick's key is pressed or released */
		CORO_WAIT(( keys = brick_keys()) != pressed );
		pressed = keys;

		switch ( pressed ) {
		/* Quit */
		case EV3_KEY_BACK:
			command = STOP;
			alive = 0;
			break;
		/* Stop */
		case EV3_KEY__NONE_:
		case EV3_KEY_CENTER:
			command = STOP;
			break;
		/* Forward */
		case EV3_KEY_UP:
			command = FORTH;
 			break;
		/* Backward */
		case EV3_KEY_DOWN:
			command = BACK;
			break;
		/* Left */
		case EV3_KEY_LEFT:
			command = LEFT;
			break;
		/* Right */
		case EV3_KEY_RIGHT:
			command = RIGHT;
			break;
		}
		CORO_YIELD();
	}
	CORO_END();
}

/* Coroutine of control the motors */
CORO_DEFINE( drive )
{
	CORO_LOCAL int speed_linear, speed_circular;
	CORO_LOCAL int state = STOP;

	CORO_BEGIN();
	speed_linear = max_speed * SPEED_LINEAR / 100;
	speed_circular = max_speed * SPEED_CIRCULAR / 100;

	for ( ; ; ) {
		/* Waiting new command */
		CORO_WAIT( state != command );

		switch ( command ) {

		case STOP:
			tacho_stop( MOTOR_BOTH );
			/* Waiting the vehicle is stopped */
			CORO_WAIT( !tacho_is_running( MOTOR_BOTH ));
			break;

		case FORTH:
			tacho_set_speed_sp( MOTOR_BOTH, -speed_linear );
			tacho_run_forever( MOTOR_BOTH );
			break;

		case BACK:
			tacho_set_speed_sp( MOTOR_BOTH, speed_linear );
			tacho_run_forever( MOTOR_BOTH );
			break;

		case LEFT:
			tacho_set_speed_sp( MOTOR_LEFT, speed_circular );
			tacho_set_speed_sp( MOTOR_RIGHT, -speed_circular );
			tacho_run_forever( MOTOR_BOTH );
			break;

		case RIGHT:
			tacho_set_speed_sp( MOTOR_LEFT, -speed_circular );
			tacho_set_speed_sp( MOTOR_RIGHT, speed_circular );
			tacho_run_forever( MOTOR_BOTH );
			break;
		}
		state = command;
	}
	CORO_END();
}

int main( void )
{
	printf( "Waiting the EV3 brick online...\n" );
	if ( !brick_init()) return ( 1 );

	printf( "*** ( EV3 ) Hello! ***\n" );
	alive = init();

	while ( alive ) {
		CORO_CALL( handle_ir_control );
		CORO_CALL( handle_brick_control );
		CORO_CALL( drive );
		sleep_ms( 10 );
	}
	brick_uninit();

	printf( "*** ( EV3 ) Bye! ***\n" );
	return ( 0 );
}

File: python/ev3dev/eg/remote.py
# -*- coding: utf-8 -*-

import sys
from ev3dev import *

# Motor speed for linear motion, in percents
SPEED_LINEAR   = 75
# ... for circular motion
SPEED_CIRCULAR = 50

# Motor maximal speed (will be detected)
max_speed = 0

MOTOR_LEFT  = OUTC
MOTOR_RIGHT = OUTB
MOTOR_BOTH  = MOTOR_LEFT | MOTOR_RIGHT

IR_CHANNEL = 0

# IR sensor port (will be detected)
ir = SOCKET__NONE_

STOP  = 0
FORTH = 1
BACK  = 2
LEFT  = 3
RIGHT = 4

# Command for `drive` coroutine
command = STOP

# Program is alive
alive = True

def init():
    global max_speed, ir

    if tacho_is_plugged( MOTOR_BOTH, TACHO_TYPE__NONE_ ):  # any type of motor
        max_speed = tacho_get_max_speed( MOTOR_LEFT )
        tacho_reset( MOTOR_BOTH )
    else:
        print '''Please, plug LEFT motor in C port,
        RIGHT motor in B port and try again.'''
        # Quit, inoperative without motors
        return False

    ir = sensor_search( LEGO_EV3_IR )
    if ir:
        ir_set_mode_ir_remote( ir )
        print '''IR remote control:
        RED UP   & BLUE UP   - forward
        RED DOWN & BLUE DOWN - backward
        RED UP   | BLUE DOWN - left
        RED DOWN | BLUE UP   - right'''
    else:
        print '''IR sensor is NOT found,
        use the EV3 brick buttons.'''

    print 'Press BACK on the EV3 brick for EXIT...'
    return True

IR_ACTIONS = {
    RED_UP_BLUE_UP     : FORTH,
    RED_DOWN_BLUE_DOWN : BACK,
    RED_UP             : LEFT,
    RED_UP_BLUE_DOWN   : LEFT,
    BLUE_DOWN          : LEFT,
    RED_DOWN           : RIGHT,
    RED_DOWN_BLUE_UP   : RIGHT,
    BLUE_UP            : RIGHT,
}

def coro_handle_ir_control():
    """
    Coroutine of IR remote control handling.
    """
    global command

    if ir == SOCKET__NONE_:
        # Do nothing if no IR sensor
        while True:
            yield

    pressed = btns = IR_REMOTE__NONE_
    while True:
        # Waiting any IR RC button is pressed or released
        while btns == pressed:
            btns = sensor_get_value( IR_CHANNEL, ir )
            yield

        pressed = btns
        # Set command for `drive` coroutine
        if pressed in IR_ACTIONS:
            command = IR_ACTIONS[ pressed ]
        else:
            command = STOP
        yield

KEY_ACTIONS = {
    EV3_KEY_UP    : FORTH,
    EV3_KEY_DOWN  : BACK,
    EV3_KEY_LEFT  : LEFT,
    EV3_KEY_RIGHT : RIGHT,
}

def coro_handle_brick_control():
    """
    Coroutine of the EV3 brick keys handling.
    """
    global command, alive

    pressed = keys = EV3_KEY__NONE_
    while True:
        # Waiting any brick's key is pressed or released
        while keys == pressed:
            keys = brick_keys()
            yield

        pressed = keys
        # Set command for `drive` coroutine
        if pressed in KEY_ACTIONS:
            command = KEY_ACTIONS[ pressed ]
        else:
            command = STOP
            if pressed == EV3_KEY_BACK:
                alive = False  # Quit
        yield

def coro_drive():
    """
    Coroutine of control the motors.
    """
    speed_linear = max_speed * SPEED_LINEAR / 100
    speed_circular = max_speed * SPEED_CIRCULAR / 100
    state = STOP

    while True:
        # Waiting new command
        while state == command:
            yield

        if command == STOP:
            tacho_stop( MOTOR_BOTH )
            # Waiting the vehicle is stopped
            while tacho_is_running( MOTOR_BOTH ):
                yield

        elif command == FORTH:
            tacho_set_speed_sp( MOTOR_BOTH, -speed_linear )
            tacho_run_forever( MOTOR_BOTH )

        elif command == BACK:
            tacho_set_speed_sp( MOTOR_BOTH, speed_linear )
            tacho_run_forever( MOTOR_BOTH )

        elif command == LEFT:
            tacho_set_speed_sp( MOTOR_LEFT, speed_circular )
            tacho_set_speed_sp( MOTOR_RIGHT, -speed_circular )
            tacho_run_forever( MOTOR_BOTH )

        elif command == RIGHT:
            tacho_set_speed_sp( MOTOR_LEFT, -speed_circular )
            tacho_set_speed_sp( MOTOR_RIGHT, speed_circular )
            tacho_run_forever( MOTOR_BOTH )

        state = command
        yield

if __name__ == '__main__':
    print 'Waiting the EV3 brick online...'
    if not brick_init(): sys.exit( 1 )

    print '*** ( EV3 ) Hello! ***'
    alive = init()

    handle_ir_control = coro_handle_ir_control()
    handle_brick_control = coro_handle_brick_control()
    drive = coro_drive()

    while alive:
        handle_ir_control.next()
        handle_brick_control.next()
        drive.next()
        sleep_ms( 10 )

    brick_uninit()
    print
    print '*** ( EV3 ) Bye! ***'
